plugin/plugin_test.go

```diff
@@ -56,117 +52,211 @@ func TestLogicOfRateLimit(t *testing.T) {
 	defer svr.Close()
 
 	requestURL := fmt.Sprintf("%v?id=1", svr.URL)
-	log.Println("requestURL", requestURL)
 
-	for i := 0; i < 1000; i++ {
-		if i%200 == 0 {
-			time.Sleep(time.Second * 1)
-		}
-		wg.Add(1)
-		go func(wg *sync.WaitGroup, mut *sync.Mutex, i int) {
-			defer wg.Done()
-			for j := 0; j < 1; j++ {
-				resp, err := svr.Client().Get(requestURL)
-				if err != nil {
-					fmt.Printf("Error: %s", err)
-					// t.Errorf("Error in %v, expected: %v, got: %v", test.url, test.expectedStatusCode, err.Error())
-				}
-				if resp.StatusCode == 200 {
-					mut.Lock()
-					results = append(results, fmt.Sprintf("PASS: i=%v, j=%v, time=%v", i, j, time.Since(initialTime).Milliseconds()))
-					mut.Unlock()
-				} else {
-					mut.Lock()
-					results = append(results, fmt.Sprintf("FAIL: i=%v, j=%v, time=%v", i, j, time.Since(initialTime).Milliseconds()))
-					mut.Unlock()
+	ticker := time.NewTicker(time.Second * 1)
+
+	j := 0
+
+	type Error struct {
+		Index int
+		Err   error
+	}
+
+	errChan := make(chan *Error, N)
+
+	for {
+		<-ticker.C
+
+		for i := 0; i < 200; i++ {
+
+			wg.Add(1)
+			go func(wg *sync.WaitGroup, i int) {
+				defer wg.Done()
+				for j := 0; j < 1; j++ {
+					resp, err := svr.Client().Get(requestURL)
+					errChan <- &Error{
```

Comment:
Is this correct, shall we send even if no error?

Reply:
I saw on the reference there were sending all the errors and checking error after receiving data from channel.
![image](https://github.com/VermaShivansh/coraza-ratelimit-plugin/assets/75882684/f7866025-0bcd-4fbc-a0f4-c421a252521a)
This way it helps to end the receiving end of the Go channel as we are not running an infinite loop but rather a finite loop of N iterations receiving Go channel.

If we send only if there is an error we will have to create a never-ending for-select and end it manually once all the iterations are done by sending a specific kind of value in the error channel. (probably nil)

What do you say should we change?

Reply:
Thanks for the explanation. Sounds good.